File: lib/real_debrid/api/streaming.ex
defmodule RealDebrid.Api.Streaming do
  @moduledoc """
  Streaming-related API functions.
  """

  alias RealDebrid.Client

  @type streaming_transcode :: %{
          apple: map(),
          dash: map(),
          live_mp4: map(),
          h264_webm: map()
        }

  @type video_detail :: %{
          stream: String.t(),
          lang: String.t(),
          lang_iso: String.t(),
          codec: String.t(),
          colorspace: String.t(),
          width: integer(),
          height: integer()
        }

  @type audio_detail :: %{
          stream: String.t(),
          lang: String.t(),
          lang_iso: String.t(),
          codec: String.t(),
          sampling: integer(),
          channels: float()
        }

  @type subtitle_detail :: %{
          stream: String.t(),
          lang: String.t(),
          lang_iso: String.t(),
          type: String.t()
        }

  @type media_infos :: %{
          filename: String.t(),
          hoster: String.t(),
          link: String.t(),
          type: String.t(),
          season: String.t() | nil,
          episode: String.t() | nil,
          year: String.t() | nil,
          duration: float(),
          bitrate: integer(),
          size: integer(),
          details: %{
            video: map(),
            audio: map(),
            subtitles: map()
          },
          poster_path: String.t(),
          audio_image: String.t(),
          backdrop_path: String.t()
        }

  @doc """
  Get transcoding options for a media file.

  ## Parameters

    - `client` - The Real-Debrid client
    - `id` - The file ID

  ## Returns

    - `{:ok, transcode}` on success
    - `{:error, reason}` on failure
  """
  @spec get_transcode(Client.t(), String.t()) :: {:ok, streaming_transcode()} | {:error, term()}
  def get_transcode(%Client{} = client, id) do
    case Client.get(client, "/streaming/transcode/#{id}") do
      {:ok, body, _headers} ->
        {:ok,
         %{
           apple: body["apple"] || %{},
           dash: body["dash"] || %{},
           live_mp4: body["liveMP4"] || %{},
           h264_webm: body["h264WebM"] || %{}
         }}

      {:error, reason} ->
        {:error, reason}
    end
  end

  @doc """
  Get media information for a file.

  ## Parameters

    - `client` - The Real-Debrid client
    - `id` - The file ID

  ## Returns

    - `{:ok, media_infos}` on success
    - `{:error, reason}` on failure
  """
  @spec get_media_infos(Client.t(), String.t()) :: {:ok, media_infos()} | {:error, term()}
  def get_media_infos(%Client{} = client, id) do
    case Client.get(client, "/streaming/mediaInfos/#{id}") do
      {:ok, body, _headers} ->
        {:ok,
         %{
           filename: body["filename"],
           hoster: body["hoster"],
           link: body["link"],
           type: body["type"],
           season: body["season"],
           episode: body["episode"],
           year: body["year"],
           duration: body["duration"],
           bitrate: body["bitrate"],
           size: body["size"],
           details: %{
             video: body["details"]["video"] || %{},
             audio: body["details"]["audio"] || %{},
             subtitles: body["details"]["subtitles"] || %{}
           },
           poster_path: body["poster_path"],
           audio_image: body["audio_image"],
           backdrop_path: body["backdrop_path"]
         }}

      {:error, reason} ->
        {:error, reason}
    end
  end
end

File: api/streaming.go
package api

import (
	"encoding/json"
	"net/http"

	real_debrid "github.com/sushydev/real_debrid_go"
)

type StreamingTranscode struct {
	Apple    map[string]string `json:"apple"`
	Dash     map[string]string `json:"dash"`
	LiveMP4  map[string]string `json:"liveMP4"`
	H264WebM map[string]string `json:"h264WebM"`
}

func GetStreamingTranscode(client *real_debrid.Client, id string) (*StreamingTranscode, error) {
	url := client.GetUrl("/streaming/transcode/" + id)
	req, err := http.NewRequest("GET", url.String(), nil)
	if err != nil {
		return nil, err
	}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if err := client.HandleResponseCode(resp, 200); err != nil {
		return nil, err
	}
	var s StreamingTranscode
	if err := json.NewDecoder(resp.Body).Decode(&s); err != nil {
		return nil, err
	}
	return &s, nil
}

type MediaInfos struct {
	Filename string  `json:"filename"`
	Hoster   string  `json:"hoster"`
	Link     string  `json:"link"`
	Type     string  `json:"type"`
	Season   *string `json:"season"`
	Episode  *string `json:"episode"`
	Year     *string `json:"year"`
	Duration float64 `json:"duration"`
	Bitrate  int     `json:"bitrate"`
	Size     int     `json:"size"`
	Details  struct {
		Video map[string]struct {
			Stream     string `json:"stream"`
			Lang       string `json:"lang"`
			LangISO    string `json:"lang_iso"`
			Codec      string `json:"codec"`
			Colorspace string `json:"colorspace"`
			Width      int    `json:"width"`
			Height     int    `json:"height"`
		} `json:"video"`
		Audio map[string]struct {
			Stream   string  `json:"stream"`
			Lang     string  `json:"lang"`
			LangISO  string  `json:"lang_iso"`
			Codec    string  `json:"codec"`
			Sampling int     `json:"sampling"`
			Channels float64 `json:"channels"`
		} `json:"audio"`
		Subtitles map[string]struct {
			Stream  string `json:"stream"`
			Lang    string `json:"lang"`
			LangISO string `json:"lang_iso"`
			Type    string `json:"type"`
		} `json:"subtitles"`
	} `json:"details"`
	PosterPath   string `json:"poster_path"`
	AudioImage   string `json:"audio_image"`
	BackdropPath string `json:"backdrop_path"`
}

func GetStreamingMediaInfos(client *real_debrid.Client, id string) (*MediaInfos, error) {
	url := client.GetUrl("/streaming/mediaInfos/" + id)
	req, err := http.NewRequest("GET", url.String(), nil)
	if err != nil {
		return nil, err
	}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if err := client.HandleResponseCode(resp, 200); err != nil {
		return nil, err
	}
	var m MediaInfos
	if err := json.NewDecoder(resp.Body).Decode(&m); err != nil {
		return nil, err
	}
	return &m, nil
}
